src/client/api/pipeline.c

```diff
@@ -0,0 +1,137 @@
+/**
+ * (C) Copyright 2015-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+#define D_LOGFAC DD_FAC(client)
+
+#include <daos/pipeline.h>
+#include <daos/task.h>
+#include "client_internal.h"
+#include "task_internal.h"
+
+void
+daos_pipeline_init(daos_pipeline_t *pipeline)
+{
+	pipeline->version          = 1;
+	pipeline->num_filters      = 0;
+	pipeline->filters          = NULL;
+	pipeline->num_aggr_filters = 0;
+	pipeline->aggr_filters     = NULL;
+}
+
+void
+daos_filter_init(daos_filter_t *filter)
+{
+	bzero((void *)filter, sizeof(daos_filter_t));
+}
+
+int
+daos_pipeline_add(daos_pipeline_t *pipeline, daos_filter_t *filter)
+{
+	daos_filter_t **ptr;
+
+	if (!strncmp((char *)filter->filter_type.iov_buf, "DAOS_FILTER_CONDITION",
+		     filter->filter_type.iov_len)) {
+		D_REALLOC_ARRAY(ptr, pipeline->filters, pipeline->num_filters,
+				pipeline->num_filters + 1);
+		if (ptr == NULL)
+			return -DER_NOMEM;
+		pipeline->filters                        = ptr;
+
+		pipeline->filters[pipeline->num_filters] = filter;
+		pipeline->num_filters += 1;
+	} else if (!strncmp((char *)filter->filter_type.iov_buf, "DAOS_FILTER_AGGREGATION",
+			    filter->filter_type.iov_len)) {
+		D_REALLOC_ARRAY(ptr, pipeline->aggr_filters, pipeline->num_aggr_filters,
+				pipeline->num_aggr_filters + 1);
+		if (ptr == NULL)
+			return -DER_NOMEM;
+		pipeline->aggr_filters                             = ptr;
+
+		pipeline->aggr_filters[pipeline->num_aggr_filters] = filter;
+		pipeline->num_aggr_filters += 1;
+	} else {
+		return -DER_INVAL;
+	}
+
+	return 0;
+}
+
+int
+daos_filter_add(daos_filter_t *filter, daos_filter_part_t *part)
+{
+	daos_filter_part_t **ptr;
+
+	D_REALLOC_ARRAY(ptr, filter->parts, filter->num_parts, filter->num_parts + 1);
+	if (ptr == NULL)
+		return -DER_NOMEM;
+
+	filter->parts                    = ptr;
+
+	filter->parts[filter->num_parts] = part;
+	filter->num_parts += 1;
+
+	return 0;
+}
+
+int
+daos_pipeline_check(daos_pipeline_t *pipeline)
+{
+	return dc_pipeline_check(pipeline);
+}
+
+static void
+free_filters(daos_filter_t **filters, uint32_t nfilters)
+{
+	uint32_t i;
+
+	for (i = 0; i < nfilters; i++) {
+		D_ASSERT(filters[i] != NULL);
+		if (filters[i]->num_parts > 0)
+			D_ASSERT(filters[i]->parts != NULL);
+		if (filters[i]->num_parts > 0)
+			D_FREE(filters[i]->parts);
+	}
+}
+
+void
+daos_pipeline_free(daos_pipeline_t *pipeline)
+{
+	D_ASSERT(pipeline != NULL);
```

Comment:
should not assert here and return DER_INVAL instead if argument is NULL.

Reply:
ok